### PR TITLE
MC/CUDA: add mc managed and tests

### PR DESCRIPTION
**Pull request by shimmybalsam**
## What
Adding support for memory type cudaManaged across UCC.

### CODE REVIEW

**Comment by shimmybalsam** [comment]
@edgargabriel please see small changes made in TL/ROCM

File: src/components/mc/cuda/mc_cuda.c [hunk 1]
```diff
@@ -139,20 +142,23 @@ static ucc_status_t ucc_mc_cuda_mem_alloc(ucc_mc_buffer_header_t **h_ptr,
     h->from_pool = 0;
     h->mt        = UCC_MEMORY_TYPE_CUDA;
```
Comment (samnordmann) [inline src/components/mc/cuda/mc_cuda.c:143]: shouldn't we use `mt` here ?

File: src/components/mc/cuda/mc_cuda.c [hunk 2]
```diff
@@ -139,20 +142,23 @@ static ucc_status_t ucc_mc_cuda_mem_alloc(ucc_mc_buffer_header_t **h_ptr,
...
 
 static ucc_status_t ucc_mc_cuda_mem_pool_alloc(ucc_mc_buffer_header_t **h_ptr,
-                                               size_t                   size)
+                                               size_t                   size,
+                                               ucc_memory_type_t        mt)
 {
     ucc_mc_buffer_header_t *h = NULL;
-    if (size <= MC_CUDA_CONFIG->mpool_elem_size) {
+    if (size <= MC_CUDA_CONFIG->mpool_elem_size &&
+        mt != UCC_MEMORY_TYPE_CUDA_MANAGED) {
```
Comment (samnordmann) [inline src/components/mc/cuda/mc_cuda.c:156]: If I understand correctly we choose not to allocate a mpool for cuda managed ?

File: src/core/ucc_coll.c [hunk 3]
```diff
@@ -219,6 +219,9 @@ UCC_CORE_PROFILE_FUNC(ucc_status_t, ucc_collective_init,
         case UCC_MEMORY_TYPE_CUDA:
```
Comment (samnordmann) [inline src/core/ucc_coll.c:219]: nit: just add the case cuda managed under the case cuda:
```
case UCC_MEMORY_TYPE_CUDA:
case UCC_MEMORY_TYPE_CUDA_MANAGED:
     coll_ee_type = UCC_EE_CUDA_STREAM;
     break;
```

File: test/gtest/core/test_mc_reduce.cc [hunk 4]
```diff
@@ -57,6 +57,9 @@ class test_mc_reduce : public testing::Test {
             }
 #endif
             break;
+        case UCC_MEMORY_TYPE_CUDA_MANAGED:
```
Comment (samnordmann) [inline test/gtest/core/test_mc_reduce.cc:60]: why do we treat Cuda managed differently than Cuda here ? What with the triggered case ?
If there is not difference then you can simply add `case UCC_MEMORY_TYPE_CUDA_MANAGED:` below `case UCC_MEMORY_TYPE_CUDA:`

**Comment by edgargabriel** [comment]
> @edgargabriel please see small changes made in TL/ROCM

@shimmybalsam thanks for the heads up. I tested your branch and things still work for ROCm with your change.

File: src/components/mc/cuda/mc_cuda.c (hunk 1 above)
Comment (shimmybalsam) [reply src/components/mc/cuda/mc_cuda.c:143]: no, there are certain cases (free) for example that we check what h->mt is and based on that choose what function to call, and besides alloc we want all other mc functions to treat cudaManages as cuda.

File: src/components/mc/cuda/mc_cuda.c (hunk 2 above)
Comment (shimmybalsam) [reply src/components/mc/cuda/mc_cuda.c:156]: correct

File: test/gtest/core/test_mc_reduce.cc (hunk 4 above)
Comment (shimmybalsam) [reply test/gtest/core/test_mc_reduce.cc:60]: we don't support triggered with cudaManaged

**Review by samnordmann — APPROVED** [review]

**Review by Sergei-Lebedev — APPROVED** [review]

File: src/components/mc/cuda/mc_cuda.c [hunk 5]
```diff
@@ -126,7 +127,9 @@ static ucc_status_t ucc_mc_cuda_mem_alloc(ucc_mc_buffer_header_t **h_ptr,
                  sizeof(ucc_mc_buffer_header_t));
         return UCC_ERR_NO_MEMORY;
     }
-    st = cudaMalloc(&h->addr, size);
+    st = (mt == UCC_MEMORY_TYPE_CUDA) ? cudaMalloc(&h->addr, size) :
+                                        cudaMallocManaged(&h->addr, size,
+                                                          cudaMemAttachGlobal);
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply src/components/mc/cuda/mc_cuda.c:132]: Needed for compilation - It is supposed to be this way by default but I think it is compiler specific

**Pull request merged**